text/0000-pod.md

```diff
@@ -0,0 +1,141 @@
+- Feature Name: pod
+- Start Date: 2015-03-04
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Split the `Copy` trait in two traits: `Pod` and `Copy`. `Pod` indicates that a
+type can be cloned by just copying its bits ("`memcpy`"), and the new `Copy`
+indicates that the type should be implicitly copied instead of moved.
+
+# Motivation
+
+### `Range` is not `Copy` (or deriving `Copy` is hard)
+
+We have a guideline that says that iterators should not be implicitly copyable
+(i.e. they shouldn't implement `Copy`). The `Range` struct, which is what
+`a..b` desugars to, is an iterator, and under this guideline it shouldn't be
+`Copy`, but this hurts ergonomics in other cases where it's not used as an
+iterator, for example a struct that contains `Range` can't be marked as `Copy`,
+because `Range` is not `Copy`.
+
+This is a more general problem, if upstream decides that some type shouldn't be
+`Copy`, even if it could, then downstream can't make `Copy`able structs/enums
+that contain said type.
+
+### `Cell` only admits `Copy` data
+
+This is unnecessarily restrictive because the data doesn't need to be
+implicitly copyable, the actual requirement is that the data should be
+cloneable by just `memcpy`ing it.
+
+### How splitting `Copy` helps with this?
+
+Implementing `Pod` for a type is much less controversial than implementing
+`Copy` on it. This should result in more types marked as `Pod` in libraries,
+which in turn gives downstream users more control about what can be `Copy` as
+the only requirement is that the fields must be `Pod`.
+
+`Cell` can be modified to have a `Pod` bound instead of a `Copy` one, this
+would allow more types, like iterators, to be stored in a `Cell`.
+
+# Detailed design
+
+### The traits
+
+A `Pod` marker trait will be added to the stdlib, and the `Copy` trait will be
+updated as follows:
+
+```
+/// Indicates that this type can be cloned by just copying its bits (`memcpy`)
+#[lang = "pod"]
+trait Pod: MarkerTrait {}
+
+/// Indicates that this type will be implicitly copied instead of moved
+#[lang = "copy"]
+trait Copy: Pod {}
+```
+
+### Compiler rules
+
+Update the compiler to enforce the following rules:
+
+- `Pod` can only be implemented by structs/enums whose fields are also `Pod`.
+- Built-in/primitive types that were automatically considered `Copy` by the
+  compiler (`i8`, `&T`, `(A, B)`, `[T; N]`, etc), will now be considered to be
+  both `Pod` and `Copy`.
+- `Copy` restrictions will carry over to `Pod`, e.g. types with destructors
+  can't implement `Pod`, nor `Copy` because of the `trait Copy: Pod`
+  requirement.
+
+### `#[derive]` syntax extensions
+
+To avoid breaking the world, the `#[derive(Copy)]` syntax extension will be
+updated to derive *both* `Copy` and `Pod`:
```

Comment:
In the case of `Eq`, if we make it an `unsafe` trait (why haven't we done it already?), the `derive` implementation of `Eq` would be incorrect unless `derive` was also supplying the `PartialEq` implementation - the same goes for `Ord` and `PartialOrd`, as well.